docs/developer-guide/firstanalysis.md

```diff
@@ -0,0 +1,144 @@
+
+  assert(x > 0);  // Thus, this assertion should hold!
+
+  return 0;
+}
+```
+
+If you run goblint out of the box on this example it will not work: `./goblint --enable dbg.debug example.c` will claim that the assertion in unknown.
+Goblint could verify that this asertion does hold using interval analysis (`--enable ana.int.interval`), but here we will implement a simple sign analysis instead.
+
+We begin by registering and running a new analysis called "signs":
+
+1. Create a copy of the unit analysis (*src/analyses/unit.ml*) into the file *src/analyses/signs.ml*.
+2. Inside the file, change the name from "unit" to "signs".
+3. Try to run goblint with the new analysis enabled: `./goblint --sets "ana.activated[+]" signs --enable dbg.debug example.c`. The result will still be that nothing is verified.
+4. Now edit the analysis to pass the test. :)
+
+
+## Designing the domain
+
+The last step might need some clarifications. We first need to design the abstract domain. It may help if you have read some theoretical tutorial on abstract domains. Our first sign lattice will simply contain the elements `{-, 0, +}` with top and bottom added. For this, we define first the signs and then we lift it. Insert the following into your file just before the module `Spec` is defined.
+
+```
+module Signs =
+struct
+  type t = Neg | Zero | Pos [@@deriving eq, to_yojson]
+  let name () = "signs"
+  let show x = match x with
+  | Neg -> "-"
+  | Pos -> "+"
+  | Zero -> "0"
+
+  (* We need the following to generate output... *)
+  include Printable.StdPolyCompare
+  include Printable.PrintSimple (struct
+    type t' = t
+    let name = name
+    let show = show
+  end)
+  let hash = Hashtbl.hash
+
+  (* Here as important domain-specific implementations,
+   * there are some mistakes for you to fix... *)
+  let of_int i =
+    if i < Int64.zero then Zero
+    else if i > Int64.zero then Zero
+    else Zero
+
+  let gt x y = match x,y with
+  | Pos, Neg | Zero, Neg -> true
+  | _ -> false
+
+end
+
+module SL =
+struct
+  include Lattice.Flat (Signs) (Printable.DefaultNames)
+  let of_int i = `Lifted (Signs.of_int i)
+
+  let gt x y = match x, y with
+  | `Lifted x, `Lifted y -> Signs.gt x y
+  | _ -> false
+end
+```
+
+There are some flawed definitions above to get you started, but let us first tell the analysis to actually use this domain.
+For that, we change the similar lines that currently use the unit domain to now use a map from variables to the newly created sign domains.
+
+```
+module D = MapDomain.MapBot (Basetype.Variables) (SL)
+module G = Lattice.Unit
+module C = D
+```
+
+We will not care about the globals and the context, but it's convenient to have the same context as the domain itself.
+
+
+## Implementing the sign analysis
+
+The key part now is to define transfer functions for assignment. Here is a simple version that will suffice for our example. Replace the definition of `assign` in your file with the following:
```

Comment:
The other default definitions in the unit analysis do undesired things like passing the map into functions.

Reply:
Mmm... again, I don't actually care, but as part of our official docs, this (and the soundness) is problematic in the sense that if people use it as a template to actually write their own analyses. My intention was more to have a throwaway analysis to get people started on goblint, not a template for how to start on a new analysis that should be used.

Reply:
The soundness issue was more serious, but these unnecessary stuff happening on features that are beyond the scope of the tutorial --- is it harmful to ignore it?

Reply:
If global variables are also excluded, then I suppose it might be fine to pass the map into the callee. Currently globals are added and could be modified in the callee, but `combine` goes back to the values of globals before the call.